contracts/apps/TokenIOU.sol

```diff
@@ -0,0 +1,214 @@
+import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
+
+import "@openzeppelin/contracts/interfaces/IERC1271.sol";
+
+contract TokenIOU is Initializable, OwnableUpgradeable, UUPSUpgradeable {
+    IERC20Upgradeable public token;
+
+    /// only a single redeem per hash
+    ///
+    /// allows us to check which username was already reserved by a profile
+    mapping(bytes32 redeemHash => uint48 time) public redeemed;
+
+    // Redeem event
+    event Redeem(address indexed from, address indexed to, uint256 amount);
+
+    function initialize(IERC20Upgradeable _token) public initializer {
+        __Ownable_init();
+        __UUPSUpgradeable_init();
+        token = _token;
+    }
+
+    /**
+        * @notice Get the hash of the redeem request
+        * @param from       The address of the redeemer
+        * @param amount     The amount of tokens to redeem
+        * @param validUntil The timestamp until which the redeem is valid
+        * @param validAfter  The timestamp from which the redeem is valid
+        * @param nonce      A unique number to prevent replay attacks
+        * @return The hash of the redeem request
+
+        Can be used to easily generate a valid hash for the redeem function
+    */
+    function getHash(
+        address from,
+        uint256 amount,
+        uint48 validUntil,
+        uint48 validAfter,
+        uint256 nonce
+    ) public view returns (bytes32) {
+        return
+            keccak256(
+                abi.encodePacked(
+                    from,
+                    amount,
+                    validUntil,
+                    validAfter,
+                    nonce,
+                    block.chainid,
+                    address(this)
+                )
+            );
+    }
+
+    /**
+     * @notice Redeem tokens
+     * @param from       The address of the redeemer
+     * @param amount     The amount of tokens to redeem
+     * @param validUntil The timestamp until which the redeem is valid
+     * @param validAfter  The timestamp from which the redeem is valid
+     * @param nonce      A unique number to prevent replay attacks
+     * @param signature  The signature of the redeem request
+     */
+    function redeem(
+        address from,
+        uint256 amount,
+        uint48 validUntil,
+        uint48 validAfter,
+        uint256 nonce,
+        bytes calldata signature
+    ) public {
+        uint48 currentTime = uint48(block.timestamp);
+        require(currentTime >= validAfter, "TokenIOU: expired or not due");
+        require(currentTime < validUntil, "TokenIOU: expired or not due");
+
+        bytes32 redeemHash = keccak256(
+            abi.encodePacked(
+                "\x19Ethereum Signed Message:\n32",
+                getHash(from, amount, validUntil, validAfter, nonce)
+            )
+        );
+
+        require(redeemed[redeemHash] == 0, "TokenIOU: Already redeemed");
+
+        // check the sender's signature
```

Comment:
No, 

// assigning msg.sender to a new variable to because I'm lazy to write msg.sender all the time

🙃

Reply:
So, it's back to being msg.sender